lib/pacemaker/pcmk_output.c

```diff
@@ -579,6 +809,241 @@ digests_xml(pcmk__output_t *out, va_list args)
     return pcmk_rc_ok;
 }
 
+#define STOP_SANITY_ASSERT(lineno) do {                                 \
+        if(current && current->details->unclean) {                      \
+            /* It will be a pseudo op */                                \
+        } else if(stop == NULL) {                                       \
+            crm_err("%s:%d: No stop action exists for %s",              \
+                    __func__, lineno, rsc->id);                         \
+            CRM_ASSERT(stop != NULL);                                   \
+        } else if (pcmk_is_set(stop->flags, pe_action_optional)) {      \
+            crm_err("%s:%d: Action %s is still optional",               \
+                    __func__, lineno, stop->uuid);                      \
+            CRM_ASSERT(!pcmk_is_set(stop->flags, pe_action_optional));  \
+        }                                                               \
+    } while(0)
+
+PCMK__OUTPUT_ARGS("rsc-action", "pe_resource_t *", "pe_node_t *", "pe_node_t *",
+                  "gboolean")
+static int
+rsc_action_default(pcmk__output_t *out, va_list args)
+{
+    pe_resource_t *rsc = va_arg(args, pe_resource_t *);
+    pe_node_t *current = va_arg(args, pe_node_t *);
+    pe_node_t *next = va_arg(args, pe_node_t *);
+    gboolean moving = va_arg(args, gboolean);
+
+    GList *possible_matches = NULL;
+    char *key = NULL;
+    int rc = pcmk_rc_no_output;
+
+    pe_node_t *start_node = NULL;
+    pe_action_t *start = NULL;
+    pe_action_t *stop = NULL;
+    pe_action_t *promote = NULL;
+    pe_action_t *demote = NULL;
+
+    rsc_action_fn_t mk_list_item = NULL;
+
+    if (!pcmk_is_set(rsc->flags, pe_rsc_managed)
+        || (current == NULL && next == NULL)) {
+        pe_rsc_info(rsc, "Leave   %s\t(%s%s)",
+                    rsc->id, role2text(rsc->role),
+                    !pcmk_is_set(rsc->flags, pe_rsc_managed)? " unmanaged" : "");
+        return rc;
+    }
+
+    if (pcmk__str_eq(out->fmt_name, "xml", pcmk__str_none)) {
+        mk_list_item = do_rsc_action_item_xml;
+    } else {
+        mk_list_item = do_rsc_action_item;
+    }
```

Comment:
It could be.  I was trying to keep the amount of code duplication to a minimum.  There's a lot of stuff in rsc_action_default that needs to be done no matter what output format, so I wanted to share all that code.  I guess that could still be done and turn do_rsc_action_item and do_rsc_action_item_xml into the custom messages.